src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs

```diff
@@ -355,6 +355,529 @@ public class b
+
+        enum E
+        {
+            x, y, z;
+        }
+    }
+}
+");
+            var secondTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    enum E
+    {
+        z, y, x;
+    }
+}
+");
+
+            var compilation = CreateCompilation(new SyntaxTree[] { firstTree });
+
+            var type1a = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.N2.N3.E");
+            Assert.True(type1a.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            compilation = compilation.ReplaceSyntaxTree(firstTree, secondTree);
+
+            var type2a = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.E");
+            Assert.True(type2a.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            // We should have the exact same set for the names.
+            Assert.Same(type1a.MergedDeclaration.Declarations[0].MemberNames, type2a.MergedDeclaration.Declarations[0].MemberNames);
+        }
+
+        [Fact]
+        public void TestMemberNamesReused_UnaffectedByDelegate2()
+        {
+            var firstTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        enum E
+        {
+            x, y, z;
+        }
+    }
+}
+");
+            var secondTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    delegate void D();
+
+    enum E
+    {
+        z, y, x;
+    }
+}
+");
+
+            var compilation = CreateCompilation(new SyntaxTree[] { firstTree });
+
+            var type1a = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.N2.N3.E");
+            Assert.True(type1a.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            compilation = compilation.ReplaceSyntaxTree(firstTree, secondTree);
+
+            var type2a = (SourceNamedTypeSymbol)compilation.GetTypeByMetadataName("N1.E");
+            Assert.True(type2a.MergedDeclaration.Declarations[0].MemberNames.Value.SetEquals(new[] { "x", "y", "z" }));
+
+            // We should have the exact same set for the names.
+            Assert.Same(type1a.MergedDeclaration.Declarations[0].MemberNames, type2a.MergedDeclaration.Declarations[0].MemberNames);
+        }
+
+        [Fact]
+        public void TestMemberNamesReused_SameTreeDifferentCompilations()
+        {
+            var firstTree = SyntaxFactory.ParseSyntaxTree(@"
+using System;
+
+namespace N1
+{
+    namespace N2.N3
+    {
+        enum E
+        {
+            x, y, z;
+        }
+    }
+}
+");
+
+            var compilation1 = CreateCompilation(new SyntaxTree[] { firstTree });
+            var compilation2 = CreateCompilation(new SyntaxTree[] { firstTree });
+
+            var type1 = (SourceNamedTypeSymbol)compilation1.GetTypeByMetadataName("N1.N2.N3.E");
+            var type2 = (SourceNamedTypeSymbol)compilation1.GetTypeByMetadataName("N1.N2.N3.E");
```

Comment:
Was `compilation2` meant to be used here?

Reply:
yup :)